src/MKL.jl

```diff
@@ -1,6 +1,41 @@
 module MKL
 
-using MKL_jll
+using Preferences
+using Libdl
+using LinearAlgebra
+
+# Choose an MKL path; taking an explicit preference as the first choice,
+# but if nothing is set as a preference, fall back to the default choice of `MKL_jll`.
+const mkl_path = something(
+    @load_preference("mkl_path", nothing),
+    "mkl_jll",
+)::String
+
+if lowercase(mkl_path) == "mkl_jll"
+    # Only load MKL_jll if we are suppoed to use it as the MKL source
+    # to avoid an unnecessary download of the (lazy) artifact.
+    import MKL_jll
+    const libmkl_rt = MKL_jll.libmkl_rt
+elseif lowercase(mkl_path) == "system"
+    # We expect the "system" MKL to already be loaded,
+    # or be on our linker search path.
+    libname = string("libmkl_rt", ".", Libdl.dlext)
+    const libmkl_rt = find_library(libname, [""])
+    libmkl_rt == "" && error("Couldn't find $libname. Try to specify the path to `libmkl_rt` explicitly.")
+else
+    # mkl_path should be a valid path to libmkl_rt.
+    const libmkl_rt = mkl_path
+    isfile(libmkl_rt) || error("Couldn't find MKL library at $libmkl_rt.")
+end
+
+# Changing the MKL provider/path preference
+function set_mkl_path(path)
+    if lowercase(path) ∉ ("mkl_jll", "system")
+        isfile(path) || error("The provided argument $path doesn't seem to be a valid path to libmkl_rt.")
+    end
+    @set_preferences!("mkl_path" => path)
+    @info("New MKL preference set; please restart Julia to see this take effect", path)
```

Comment:
Strictly speaking, you could run the code in `__init__()` directly here; wrap it up in a function to load the library, then do the LBT forwarding.  You don't need to restart Julia to get the effect.  The only problem you'd have is that things like `mklnorm` wouldn't work right, but there aren't many `ccall()`'s in this package from what I can tell.

Reply:
If we do this, I think we should make `mkl_path` and `mkl_found` non-const (either typed globals or `Ref`s) so that we can have LBT forwarding in sync with their values. In `set_mkl_path`, I would then not just forward to the new MKL but also update `mkl_path` and `mkl_found`. This way, the `ccall`s should also work fine. Does that sound good to you?

Reply:
ccall only works with consts because it only does the lookup once.

Reply:
Alright, but even if we ignore the `ccall`s, wouldn't it be misleading if we are forwarding to MKL library A while all the constants point to MKL library B? So my question about whether we should make those global constants non-constant still stands.

Reply:
A few more things:
* The `MKL_Set_Interface_Layer` and `MKL_Set_Threading_Layer` ccalls also don't work anymore (which might or might not be an issue?). 
* We currently `import MKL_jll` only if necessary (to avoid an unnecessary download of the artifact). We can't really do this nicely from within a function (specifically, `set_mkl_path`).

While I think it would be nice to allow switching the MKL library at runtime, I'm currently leaving things as is for simplicity. After all, that isn't the focus of this PR.

Reply:
@staticfloat, if you're fine with postponing runtime-switching to a different PR, than please mark this conversation as resolved.